src/server/dflycmd.h

```diff
@@ -29,29 +31,83 @@ namespace journal {
 class Journal;
 }  // namespace journal
 
+// DflyCmd is responsible for managing replication. A master instance can be connected
+// to many replica instances, what is more, each of them can open multiple connections.
+// This is why its important to understand replica lifecycle management before making
+// any crucial changes.
+//
+// A SyncInfo instance is responsible for managing a replica's state and is accessible by its
+// sync_id. Each per-thread connection is called a Flow and is represented by the FlowInfo
+// instace, accessible by its index.
+//
+// An important aspect is synchronization and efficient locking. Two levels of locking are used:
+//  1. Global locking.
+//    Member  mutex `mu_` is used for synchronizing operations connected with internal data
+//    structures.
+//  2. Per-replica locking
+//    SyncInfo contains a separate mutex that is used for replica-only routines. It is held during
+//    state transitions (start full sync, start stable state sync), cancellation and mebmer access.
+//
+// Upon first connection from the replica, a new SyncInfo is created.
+// It tranistions through the following phases:
+//  1. Preparation
+//    During this start phase the "flows" are set up - one connection for every master thread. Those
+//    connections registered by the FLOW command sent from each newly opened connection.
+//  2. Full sync
+//    This phase is initiated by the SYNC command. It makes sure all flows are connected and the
+//    replica is in a valid state.
+//  3. Stable state sync
+//    After the replica has received confirmation, that each flow is ready to transition, it sends a
+//    STARTSTABLE command. This transitions the replica into streaming journal changes.
+//  4. Cancellation
```

Comment:
i think we should discuss more the cancellation flow. there is too much specific code that is dedicated to it.

in AcceptServer I succeeded to build a flow in such way that the normal shutdown sequence and cancel/break sequence are the same. Specifically, all the shutdown and cleanup code is the same. I did it by "breaking" the client connections and then let the shutdown flow do the rest. here it's too fragmented.

Reply:
The cancellation flow is the same as the normal shutdown, it does no error specific tasks. The main goal is just to stop all processes, which can be done only thorough cancellation

CancelSyncSession will be invoked either by the error handler or by the disconnect routine (didn't plan it yet).